### PR TITLE
Implement `hubconnection.on` and invoke messages

### PR DESCRIPTION
**Pull request by zackliu**

### CODE REVIEW

File: Sources/SignalRClient/HubConnection.swift [hunk 1]
```diff
@@ -339,6 +406,92 @@ public actor HubConnection {
...
+        func create() async -> (String, TaskCompletionSource<CompletionMessage>) {
+            let newId = id + 1
+            id = newId
+            let tcs = TaskCompletionSource<CompletionMessage>()
+            invocations[String(newId)] = tcs
+            return (String(newId), tcs)
+        }
+
+        func setResult(message: CompletionMessage) async {
+            if let tcs = invocations[message.invocationId!] {
```
Comment (bjqian) [inline Sources/SignalRClient/HubConnection.swift:478]: Looks the `invocationId` of `CompletionMessage` should not be optional

File: Sources/SignalRClient/InvocationHandler.swift [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import Foundation
 
-protocol InvocationBinder {
-    func GetReturnType(invocationId: String) -> Any.Type?
-    func GetParameterTypes(methodName: String) -> [Any.Type]
-    func GetStreamItemType(streamId: String) -> Any.Type?
+protocol InvocationBinder: Sendable{
```
Comment (bjqian) [inline Sources/SignalRClient/InvocationHandler.swift:3]: The fileName `InvocationHandler` doesn't match the protocol `InvocationBinder`

File: Sources/SignalRClient/HubConnection.swift [hunk 3]
```diff
@@ -339,6 +406,92 @@ public actor HubConnection {
...
+
+        func getStreamItemType(streamId: String) -> (any Any.Type)? {
+            lock.wait()
+            defer {lock.signal()}
+            return nil // not implemented
+        }   
+    }
+
+    private actor InvocationHandler {
+        private let lock = DispatchSemaphore(value: 1)
```
Comment (bjqian) [inline Sources/SignalRClient/HubConnection.swift:465]: The `lock` is not used

File: Sources/SignalRClient/HubConnection.swift [hunk 4]
```diff
@@ -339,6 +406,92 @@ public actor HubConnection {
...
+        private var invocations: [String: TaskCompletionSource<CompletionMessage>] = [:]
+        private var id = 0
+
+        func create() async -> (String, TaskCompletionSource<CompletionMessage>) {
+            let newId = id + 1
+            id = newId
```
Comment (bjqian) [inline Sources/SignalRClient/HubConnection.swift:471]: The var `newId` looks not necessary

File: Sources/SignalRClient/HubConnection.swift [hunk 5]
```diff
@@ -90,20 +94,35 @@ public actor HubConnection {
...
-        // Invoke a method
-        return ""
+    public func invoke<TReturn>(method: String, arguments: Any...) async throws -> TReturn {
+        let (invocationId, tcs) = await invocationHandler.create()
+        invocationBinder.registerInvocation(invocationId: invocationId, types: TReturn.self)
+        defer {invocationBinder.removeInvocation(invocationId: invocationId)}
+        let invocationMessage = InvocationMessage(target: method, arguments: AnyEncodableArray(arguments), streamIds: nil, headers: nil, invocationId: invocationId)
+        let data = try hubProtocol.writeMessage(message: invocationMessage)
+        try await sendMessageInternal(data)
+        return try await tcs.task().result.value as! TReturn
```
Comment (bjqian) [inline Sources/SignalRClient/HubConnection.swift:117]: Looks the value might be `nil` when the customer invoke this method on `void` hub method. Then the type assertion is a fatal error.

File: Sources/SignalRClient/HubConnection.swift (hunk 1 above)
Comment (zackliu) [reply Sources/SignalRClient/HubConnection.swift:478]: Yes, just because they inherit from HubInvocationMessage whose invocationId is nullable

**Review by bjqian — APPROVED** [review]

**Pull request merged**